test/torchaudio_unittest/io/stream_reader_test.py

```diff
@@ -418,22 +418,82 @@ def test_stream_smoke_test(self):
         s = StreamReader(self.get_src())
         with self.assertRaises(RuntimeError):
             s.seek(-1.0)
 
+    def test_seek_invalid_mode(self):
+        """Calling `seek` with an invalid model should raise an exception"""
+        s = StreamReader(self.get_src())
+        with self.assertRaises(ValueError):
+            s.seek(10, "magic_seek")
+
+    @parameterized.expand(
+        [
+            # Test keyframe seek
+            # The source video has two key frames the first frame and 203rd frame at 8.08 second.
+            # If the seek time stamp is smaller than 8.08, it will seek into the first frame at 0.0 second.
+            ("nasa_13013.mp4", "key", 0.2, (0, 0)),
+            ("nasa_13013.mp4", "key", 8.04, (0, 0)),
+            ("nasa_13013.mp4", "key", 8.08, (0, 202)),
+            ("nasa_13013.mp4", "key", 8.12, (0, 202)),
+            ("v_SoccerJuggling_g23_c01.avi", "key", 0.0, (0, 0)),
+            ("v_SoccerJuggling_g23_c01.avi", "key", 0.3, (0, 0)),
+            ("v_SoccerJuggling_g23_c01.avi", "key", 0.5, (0, 12)),
+            # Test precise seek
+            ("nasa_13013.mp4", "precise", 0.0, (0, 0)),
+            ("nasa_13013.mp4", "precise", 0.2, (0, 5)),
+            ("nasa_13013.mp4", "precise", 8.04, (0, 201)),
+            ("nasa_13013.mp4", "precise", 8.08, (0, 202)),
+            ("nasa_13013.mp4", "precise", 8.12, (0, 203)),
+            ("v_SoccerJuggling_g23_c01.avi", "precise", 0.0, (0, 0)),
+            ("v_SoccerJuggling_g23_c01.avi", "precise", 0.3, (0, 9)),
+            ("v_SoccerJuggling_g23_c01.avi", "precise", 0.5, (0, 15)),
+            # Test any seek
+            # The source video has one keyframe every twelve frames 0, 12, 24,.. or every 0.4004 seconds.
+            ("v_SoccerJuggling_g23_c01.avi", "any", 0.0, (0, 0)),
+            ("v_SoccerJuggling_g23_c01.avi", "any", 0.133467, (0, 4), True),
+            ("v_SoccerJuggling_g23_c01.avi", "any", 0.4004, (0, 12)),
+            ("v_SoccerJuggling_g23_c01.avi", "any", 0.567233, (0, 17), True),
+            ("v_SoccerJuggling_g23_c01.avi", "any", 0.8008, (0, 24)),
+            ("v_SoccerJuggling_g23_c01.avi", "any", 1.2012, (0, 36)),
+            ("v_SoccerJuggling_g23_c01.avi", "any", 7.6076, (0, 228)),
+        ]
+    )
+    def test_seek_modes(self, src, mode, seek_time, ref_indices, assert_different=False):
+        """Testing expected behaviour for the different kinds of seek"""
+        # Using the first video stream (which is not default video stream)
+        stream_index = 0
+        # Decode all frames for reference
+        src_bin = self.get_src(src)
+        s = StreamReader(src_bin)
+        s.add_basic_video_stream(-1, stream_index=stream_index)
+        s.process_all_packets()
+        (ref_frames,) = s.pop_chunks()
+
+        s.seek(seek_time, mode=mode)
+        s.process_all_packets()
+        (frame,) = s.pop_chunks()
+
+        hyp_index, ref_index = ref_indices
+        if assert_different:
```

Comment:
I don't think this `assert_different` branch is necessary.
For "any" seek cases that result in starting with delta frames, setting `hyp_index` to a number of frames to hit the first key frame (that is a relative number of frames) should do.